Some clarification of language in .NET naming rules topic (#28237)
dotnet/docs: docs/fundamentals/code-analysis/style-rules/naming-rules.md

---
# Code-style naming rules

> In your `.editorconfig` file, you can define naming conventions for your .NET programming language code elements&mdash;such as classes, properties, and methods&mdash;and how the compiler or IDE should enforce those conventions. For example, you could specify that a public member that isn't capitalized should be treated as a compiler error, or that if a private field doesn't begin with an `_` a build warning should be issued.

Reviewer: ```suggestion
In your `.editorconfig` file, you can define naming conventions for your .NET programming language code elements&mdash;such as classes, properties, and methods&mdash;and how the compiler or IDE should enforce those conventions. For example, you could specify that a public member that isn't capitalized should be treated as a compiler error, or that if a private field doesn't begin with an `_`, a build warning should be issued.
```

Author: Does it make sense to use `*` `*` which is rendered as italics? "`.editorconfig`" is the literal file name.

Specifically, you can define a **naming rule**, which consists of three parts:

* The **symbol group** that the rule applies to, for example, public members or private fields.
* The **naming style** to associate with the rule, for example, that the name must be capitalized or start with an underscore.
* The severity level of the message when code elements included in the symbol group don't follow the naming style.

## General syntax

To define any of the above entities&mdash;a naming rule, symbol group, or naming style&mdash;set one or more properties using the following syntax:

```ini
<kind>.<entityName>.<propertyName> = <propertyValue>
```

> Property settings are grouped into entities based on their  shared `kind` and `entityName`.

Reviewer: ```suggestion
Property settings are grouped into entities based on their shared `kind` and `entityName`.
```

Reviewer: I'm not sure that this sentence will be easily understood.

Author: I'm open to other formulations. The main point I'm trying to get across here is that unlike say C# classes or XML elements, there's no special syntax for defining the parent entity; it's composed only of the various property settings.

Author: This is saying the same thing as the previous paragraph:

> To define any of the above entities—a naming rule, symbol group, or naming style—set one or more properties using the following syntax:

This paragraph doesn't add anything more; and it can be removed.

Author: Is this better?

> All the property settings for a given `kind` and `entityName` make up that specific entity definition.

Each property should only be set once, but some settings allow multiple, comma-separated values.

The order of the properties is not important.

## \<kind> values

> **\<kind>** specifies which kind of element is being defined&mdash;naming rule, symbol group, or naming style&mdash;and must be one of the following:

Reviewer: ```suggestion
**\<kind>** specifies which kind of entity is being defined&mdash;naming rule, symbol group, or naming style&mdash;and must be one of the following:
```

> Each type of definition&mdash;[naming rule](#naming-rule-properties), [symbol group](#symbol-group-properties), or [naming style](#naming-style-properties)&mdash;has its own supported properties, as described in the following sections.

Reviewer: ```suggestion
Each type of entity&mdash;[naming rule](#naming-rule-properties), [symbol group](#symbol-group-properties), or [naming style](#naming-style-properties)&mdash;has its own supported properties, as described in the following sections.
```

dotnet_naming_symbols.types.applicable_accessibilities = public, internal, private, protected, protected_internal, private_protected
```

## \<propertyName> and \<propertyValue>

### Naming rule properties

All naming rule properties are required for a rule to take effect.

| Property | Description |
| -- | -- |
| `symbols` | The name of a symbol group; the naming rule will be applied to the symbols in this group |

> | `style` | The name of the naming style which should be associated with this rule |

Author: "Previously defined in the file" isn't quite accurate, because AFAICT both the style and the symbol group could be defined later in the file. Would it be useful to rearrange the order: put first the symbol group properties, then the naming style properties, and finally the naming rule properties?

> | `style` | The name of the naming style which should be associated with this rule |

Author: Adding to the confusion: in theory you could have another editorconfig in a parent folder defining the symbol group or naming style; it wouldn't be in the same file at all. Arguably something of an edge case, but maybe just "defined elsewhere".

| `severity` |  Sets the severity with which to enforce the naming rule. Set the associated value to one of the available [severity levels](../configuration-options.md#severity-level).<sup>1</sup> |

**Notes:**

1. Severity specification within a naming rule is only respected inside development IDEs, such as Visual Studio. This setting is not understood by the C# or VB compilers, hence not respected during build. To enforce naming style rules on build, you should instead set the severity by using [code rule severity configuration](#rule-id-ide1006-naming-rule-violation). For more information, see this [GitHub issue](https://github.com/dotnet/roslyn/issues/44201).

> ### Symbol group properties

Author: I don't see any way of working around this. I think the hierarchy has to be like this:

* General syntax
  * _kind_ values
  * _entityName_
  * _propertyName_ and _propertyValue_
    * symbol group properties + values
    * naming style properties + values
    * naming rule properties + values

etc.

dotnet_naming_symbols.types.applicable_accessibilities = public, internal, private, protected, protected_internal, private_protected
```

> ## \<propertyName> and \<propertyValue>

Reviewer: ```suggestion
## \<propertyName> and \<propertyValue>

(Can you add some intro here?)
```

Author: I've done so in https://github.com/dotnet/docs/pull/28237/commits/379839b36f8802c7dc94ab5285240d25cca1c7be

The severity value must be `warning` or `error` to be [enforced on build](../overview.md#code-style-analysis). For all possible severity values, see [severity level](../configuration-options.md#severity-level).

## Example: Public member capitalization

> The following `.editorconfig` file contains a naming convention that specifies that public properties, methods, fields, events, and delegates must be capitalized. Notice that this naming convention specifies multiple kinds of symbol to apply the rule to, using a comma to separate the values.

Reviewer: We usually italicize file names.

```suggestion
The following *.editorconfig* file contains a naming convention that specifies that public properties, methods, fields, events, and delegates must be capitalized. Notice that this naming convention specifies multiple kinds of symbol to apply the rule to, using a comma to separate the values.
```

dotnet_naming_symbols.public_symbols.applicable_accessibilities = public
dotnet_naming_symbols.public_symbols.required_modifiers         = readonly

# Defining the 'first_word_upper_case_style' naming style
dotnet_naming_style.first_word_upper_case_style.capitalization = first_word_upper

# Defining the 'public_members_must_be_capitalized' naming rule, by setting the
# symbol group to the 'public symbols' symbol group,
dotnet_naming_rule.public_members_must_be_capitalized.symbols  = public_symbols
# setting the naming style to the 'first_word_upper_case_style' naming style,
dotnet_naming_rule.public_members_must_be_capitalized.style    = first_word_upper_case_style
# and setting the severity.
dotnet_naming_rule.public_members_must_be_capitalized.severity = suggestion
```

## Example: Private instance fields with underscore

> This `.editorconfig` file snippet enforces that private instance fields should start with an `_`; if that convention is not followed, the IDE will treat it as a compiler error. Private static fields will be ignored.

Reviewer: ```suggestion
This *.editorconfig* file snippet enforces that private instance fields should start with an `_`; if that convention is not followed, the IDE will treat it as a compiler error. Private static fields are ignored.
```

> Because you can only define a symbol group based on the identifiers it has (e.g. `static` or `readonly`), and not by the identifiers it doesn't have (e.g. an instance field because it doesn't have `static`), you need to define two naming rules:

Reviewer: ```suggestion
Because you can only define a symbol group based on the identifiers it has (for example, `static` or `readonly`), and not by the identifiers it doesn't have (for example, an instance field because it doesn't have `static`), you need to define two naming rules:
```

1. All private fields&mdash; `static` or not&mdash;should have the `underscored` naming style applied to them as a compiler `error`.

> 1. Private fields with `static` should have the `underscore` naming style applied to them with a severity level of `none`; in other words, ignore this case.

Reviewer: ```suggestion
1. Private fields with `static` should have the `underscored` naming style applied to them with a severity level of `none`; in other words, ignore this case.
```

```ini
[*.{cs,vb}]

# Define the 'private_fields' symbol group:
dotnet_naming_symbols.private_fields.applicable_kinds = field
dotnet_naming_symbols.private_fields.applicable_accessibilities = private

# Define the 'private_static_fields' symbol group
dotnet_naming_symbols.private_static_fields.applicable_kinds = field
dotnet_naming_symbols.private_static_fields.applicable_accessibilities = private
dotnet_naming_symbols.private_static_fields.required_modifiers = static

# Define the 'underscored' naming style
dotnet_naming_style.underscored.capitalization = pascal_case
dotnet_naming_style.underscored.required_prefix = _

# Define the 'private_fields_underscored' naming rule
dotnet_naming_rule.private_fields_underscored.symbols = private_fields
dotnet_naming_rule.private_fields_underscored.style = underscored
dotnet_naming_rule.private_fields_underscored.severity = error

> # Define the 'private_fields_underscored' naming rule

Reviewer: ```suggestion
# Define the 'private_static_fields_none' naming rule
```

dotnet_naming_rule.private_static_fields_none.symbols = private_static_fields
dotnet_naming_rule.private_static_fields_none.style = underscored
dotnet_naming_rule.private_static_fields_none.severity = none
```

> This example also demonstrates that entity definitions can be reused&mdash;the `underscored` naming style is used by both the `private_fields_underscored` and `private_static_fields_none` naming rules.

Reviewer: ```suggestion
This example also demonstrates that entity definitions can be reused. The `underscored` naming style is used by both the `private_fields_underscored` and `private_static_fields_none` naming rules.
```
